crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -550,7 +622,7 @@ impl<'w, 's, T: Resource> SystemParamFetch<'w, 's> for ResMutState<T> {
 #[doc(hidden)]
 pub struct OptionResMutState<T>(ResMutState<T>);
 
-impl<'a, T: Resource> SystemParam for Option<ResMut<'a, T>> {
+impl<'a, T: Resource> OptionalSystemParam for ResMut<'a, T> {
     type Fetch = OptionResMutState<T>;
```

Comment:
Ideally we should be able to have a generic fetch for this that handles potentially fallible fetches based on the underlying SystemParam.

As it stands this doesn't change the status quo significantly and does not net us that much of a code quality change.

Reply:
Hmm, that's tough.

So, I tried adding `impl<T: SystemParamState> SystemParamState for Option<T>`, eliminating the need for a separate `SystemParamState`.

However, this means that you have to implement `SystemParamFetch` for `Option<T>`, which breaks orphan rules, which kinda defeats the purpose of this PR.

Reply:
Hmm, maybe I can implement `SystemParamFetch` for `Option<T>`?

I'll have to do some more experiments.

Reply:
> As it stands this doesn't change the status quo significantly and does not net us that much of a code quality change.

This PR isn't about code quality for builtin types, it's about allowing users to do something that was not possible before. Without this PR, you can't (nicely) implement optional system params outside of `bevy_ecs`.

Reply:
@JoJoJet Yeah, but I can still make it better. I found a solution that works, just got to let the CI finish before I add it here.